### PR TITLE
Istiod agent

### PR DESCRIPTION
**Pull request by costinm**
Initial merging of (minimal) CA functionality from Citadel into istiod - just the GRPC cert signing and self-signed cert.

Initial merging of SDS grpc into pilot-agent - just minimal stuff to call istiod or google ca to get certs.

Tested manually - still working on running the e2e tests against this.

### CODE REVIEW

**Comment by costinm** [comment]
Note - this is built on top of https://github.com/istio/istio/pull/18132, waiting for that to merge - the last commits are relevant to CA.

File: cmd/istiod/istiod.go
Comment (pitlv2109) [inline cmd/istiod/istiod.go]: Should some of these be configurable?

File: pilot/cmd/pilot-agent/main.go
Comment (pitlv2109) [inline pilot/cmd/pilot-agent/main.go]: You can probably call out the supported version. K8s supports trustworthy jwt from 1.12 (but 1.12 didn't support audiences different than api server audiences - this came in 1.13)

File: pkg/istio-agent/sds-agent.go
Comment (pitlv2109) [inline pkg/istio-agent/sds-agent.go]: Can we start using trust domain value from the MeshConfig?

File: security/pkg/server/ca/server.go
Comment (pitlv2109) [inline security/pkg/server/ca/server.go]: s/New/NewIstiod

File: cmd/istiod/istiod.go
Comment (costinm) [reply cmd/istiod/istiod.go]: Yes, at least istio-namespace will be derived from mesh.yaml, but first priority is getting things working, and avoiding making things configurable unless there is strong enough need.

Even for istio-system - I think 1.6 istiod would be fine if we only support the common case, since we'll still ship Citadel.

File: pilot/cmd/pilot-agent/main.go
Comment (costinm) [reply pilot/cmd/pilot-agent/main.go]: My understanding is that the apiserver verification didn't support the audience - but issuing and checking using standard JWT library would still work, they are standard RS256 JWT tokens, singed with the K8S CA. 

I'm planning to add some code to support checking the tokens without a call to apiserver, so istiod can run outside of the cluster, I suspect this might take care of 1.12 as well.

File: pkg/istio-agent/sds-agent.go
Comment (costinm) [reply pkg/istio-agent/sds-agent.go]: Not here - agent has no direct access to mesh config. Injector has access to mesh config, and creates the env variables in the agent.

**Comment by howardjohn** [comment]
SDS tests failing to start proxy, you can see logs in https://storage.googleapis.com/istio-prow/pr-logs/pull/istio_istio/18314/integ-security-k8s-tests_istio/3229/artifacts/sds-citadel-flow-control-plane-auth-disabled-test-/_suite_context/istio-state-688842085/istio-ingressgateway-84c6f49f76-wht9g-istio-proxy.log. Looks like gateway needs permission to read Secret

File: pkg/istio-agent/sds-agent.go
Comment (howardjohn) [inline pkg/istio-agent/sds-agent.go]: `err` is not handled here if `fail == false` which will probably break L233

File: pkg/istio-agent/sds-agent.go
Comment (howardjohn) [inline pkg/istio-agent/sds-agent.go]: set namespace here?

**Review by howardjohn — COMMENTED** [review]
most of this requires security review, but generally lgtm

File: security/pkg/server/ca/server.go
Comment (JimmyCYJ) [inline security/pkg/server/ca/server.go]: Restore the indent here?

File: security/pkg/server/ca/server.go
Comment (JimmyCYJ) [inline security/pkg/server/ca/server.go]: Restore the indent here?

File: security/pkg/server/ca/server.go
Comment (JimmyCYJ) [reply security/pkg/server/ca/server.go]: NewIstiod is not accurate here, this method creates a new ca server.

**Comment by JimmyCYJ** [comment]
@myidpt Could you please take a look?

File: pkg/istio-agent/sds-agent.go
Comment (JimmyCYJ) [inline pkg/istio-agent/sds-agent.go]: At line 180, we trim the prefix "unix:" so that ac.SDSAddress have "unix:", but here we set ac.SDSAddress with prefix "unix:". Could you explain the reason of adding this prefix?

File: pkg/istiod/common.go
Comment (nmittler) [inline pkg/istiod/common.go]: Can you raise and issue and reference it with a TODO here?

File: pkg/istiod/istio_ca.go
Comment (nmittler) [inline pkg/istiod/istio_ca.go]: I'm not sure I understand the first sentence, maybe I'm misreading it. What does "install complexity" mean here?

File: pilot/cmd/pilot-agent/main.go
Comment (costinm) [reply pilot/cmd/pilot-agent/main.go]: It is Pilot address - in istiod the gRPC port is also shared by Citadel and Galley, so really it's istiod control plane address. Injected into pod by istiod.

Part of moving to SDS and Chiron we're removing the 'controlPlaneAuthPolicy' - instead we'll rely on the port number to detect which kind of TLS is used: 15010 plain text, 15011 Spiffee certs, 15012 - k8s-signed certs and 443 public DNS certs. We may provide some overrides - if there is strong enough justification.

File: pkg/istio-agent/sds-agent.go
Comment (costinm) [reply pkg/istio-agent/sds-agent.go]: Done

File: cmd/istiod/istiod.go
Comment (costinm) [reply cmd/istiod/istiod.go]: done

File: pilot/cmd/pilot-agent/main.go
Comment (costinm) [reply pilot/cmd/pilot-agent/main.go]: The logic is replaced by the new code - it should preserve the same semantics.

File: pkg/istiod/istio_ca.go
Comment (costinm) [reply pkg/istiod/istio_ca.go]: Added a comment. It should not be nil - like other params if it is nil things will break, no need to check each param for nil in every method.

In this case nothing wrong will happen - the CA code checks if nil and creates its own grpc server.

File: pkg/istiod/istio_ca.go
Comment (costinm) [reply pkg/istiod/istio_ca.go]: Most of the code we have is using this style, and few people know what pki is. Plus this is not citadel but istiod :-)

File: security/pkg/server/ca/server.go
Comment (costinm) [reply security/pkg/server/ca/server.go]: yes, make format-go will take care of all this. Cut&paste...

File: pkg/istio-agent/sds-agent.go
Comment (costinm) [reply pkg/istio-agent/sds-agent.go]: Envoy config expects unix:
The LocalSDS var is the file we check - no point in splitting it.

ac.SDSAddress currently has unix: for backward compat - that's how the current injector works.
The code is expected to work with the old injector so tests can pass...

File: pkg/istiod/common.go
Comment (costinm) [reply pkg/istiod/common.go]: Added it to the 'action items' doc - if anyone wants to work on it they can create an issue as well, for now we have a workaround.

File: pkg/istiod/istio_ca.go
Comment (costinm) [reply pkg/istiod/istio_ca.go]: Setting up SDS and certs involved a lot of options and steps. In istiod we only support SDS (no Secrets), and most of the steps are no longer needed.

File: pkg/istio-agent/sds-agent.go
Comment (howardjohn) [reply pkg/istio-agent/sds-agent.go]: Probably should not make another port, we already have 15020 as the status port for pilot-agent

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (howardjohn) [inline security/pkg/nodeagent/sds/sdsservice.go]: not sure how this passed CI. Can you run `BUILD_WITH_CONTAINER=1 make format`, or next time someone touches this file these indent changes will be reverted

**Review by howardjohn — COMMENTED** [review]
I am probably not qualified to review the security changes, but the rest lgtm

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (costinm) [reply security/pkg/nodeagent/sds/sdsservice.go]: I did run make format-go - and it changed 1000s of files. This is using golang built-in format.

I'm really not sure why standard go fmt is not good enough here... Buildwithcontainer didn't work last time I tried either.

**Comment by howardjohn** [comment]
We can change how we format files, but until then we should follow the way
we do now or they will just be painful for others. If it changed a lot of
stuff you probably have a different version than is run in CI, which is why
i suggested the build with container

File: pilot/cmd/pilot-agent/main.go
Comment (JimmyCYJ) [reply pilot/cmd/pilot-agent/main.go]: Thanks for the clarification.

File: pkg/istio-agent/sds-agent.go
Comment (JimmyCYJ) [reply pkg/istio-agent/sds-agent.go]: Perhaps we need to update port list used by Istio when Istiod is ready. https://istio.io/docs/setup/additional-setup/requirements/#ports-used-by-istio
And make sure there are no conflicts. According to that link, 15020 is used by Ingress gateway for Pilot health checks.

File: pilot/cmd/pilot-agent/main.go
Comment (JimmyCYJ) [inline pilot/cmd/pilot-agent/main.go]: Remove extra empty line?

**Review by JimmyCYJ — COMMENTED** [review]
@myidpt @rlenglet Could you please take a look? The changes look good to me overall.

File: pkg/istio-agent/sds-agent.go
Comment (myidpt) [inline pkg/istio-agent/sds-agent.go]: Sorry, what does this mean?

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (myidpt) [inline security/pkg/nodeagent/sds/sdsservice.go]: We should only do this in side-car mode. Falling back to the local JWT in DaemonSet mode will cause the node agent to send its own JWT to the CA.

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (myidpt) [inline security/pkg/nodeagent/sds/sdsservice.go]: Same here. Falling back to local JWT should only happen when the agent is running in sidecar mode.

File: pkg/istiod/pilotstart.go
Comment (myidpt) [reply pkg/istiod/pilotstart.go]: I think the reason is we want to differentiate the two paths. If so, can we make the path name saying the property, like "unix:/var/run/local_sds/sds_path"?

Comment (costinm) [reply pkg/istiod/pilotstart.go]: The rootfs is read only - with the exception of /etc/istio/proxy, which is a mem-mounted writable dir.

And having 'path' in the name doesn't make sense - it is a path already. I can use lower case sds if it helps. 

Another option that is now possible is to use the abstract network namespace instead of files, but not sure it's worth it.

File: pkg/istiod/istio_ca.go
Comment (costinm) [reply pkg/istiod/istio_ca.go]: Not sure how - it needs to modify iss/aud. I can rename it.

File: pkg/istiod/istio_ca.go
Comment (costinm) [reply pkg/istiod/istio_ca.go]: I've tried to reuse existing code to minimize disruption - but for newly written code I would rather use the top level pkg, in particular since this will also likely be used to authenticate the request to the XDS and MCP servers ( i.e. the pilot and galley portions of the code).

File: pilot/cmd/pilot-agent/main.go
Comment (costinm) [reply pilot/cmd/pilot-agent/main.go]: That's what I thought too - but it doesn't work. See the comment above - if watcher is disabled envoy won't start. 

The whole startup will need to be simplified and cleaned up after the switch to SDS is complete - but not in this PR.

File: pkg/istio-agent/sds-agent.go
Comment (costinm) [reply pkg/istio-agent/sds-agent.go]: No clue - cut and pasted from node agent (I think it means that same options are used when in gateway mode)

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (costinm) [reply security/pkg/nodeagent/sds/sdsservice.go]: Good point ! Will fix it ASAP.

File: pkg/istiod/istio_ca.go
Comment (myidpt) [reply pkg/istiod/istio_ca.go]: Yes, renaming it helps.

File: pkg/istio-agent/sds-agent.go
Comment (myidpt) [reply pkg/istio-agent/sds-agent.go]: Jimmy, I think we should reuse 15020 if it's already used?

**Review by myidpt — APPROVED** [review]

**Comment by lei-tang** [comment]
How to test this PR manually?

File: security/pkg/nodeagent/secretfetcher/secretfetcher.go
Comment (lei-tang) [inline security/pkg/nodeagent/secretfetcher/secretfetcher.go]: The function name in the comment does not match the actual function name.

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (lei-tang) [inline security/pkg/nodeagent/sds/sdsservice.go]: "./var/run/secrets/tokens/istio-token" is defined in the constant variables JWTPath and trustworthyJWTPath. Instead of hardcoding the path in the ReadFile(), please refer a constant variable. In addition, the same path should not be defined in multiple constant variables.

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (lei-tang) [inline security/pkg/nodeagent/sds/sdsservice.go]: I am uncertain why the localJWT option is needed. Without the localJWT option, Envoy reads the token from a file and passes it to istio agent in the context. Withou localJWT option, istio agent reads the token from a file.
If the localJWT option is introduced to improve the performance, as the token still needs to be read, there is no performance gain.

Comment (costinm) [reply security/pkg/nodeagent/sds/sdsservice.go]: This is to simplify the code - there is no need for Pilot to generate a complicated config, with the file path - in each cluster/listener - when the agent has access to the file as well.

The option is only used short term, until node-agent is still in use - once it's gone there will be no passing of tokens from envoy to SDS.

Comment (costinm) [reply security/pkg/nodeagent/sds/sdsservice.go]: As a fun fact - if the file is not found, envoy is crashing hard. 

Not passing the file from envoy is not only simpler - it also allows us the option to use Envoy GRPC library for envoy to local SDS communication, instead of using 2 grpc stacks at the same time.

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (costinm) [reply security/pkg/nodeagent/sds/sdsservice.go]: Missed this, will replace it with the constant.

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (kyessenov) [reply security/pkg/nodeagent/sds/sdsservice.go]: Be a good citizen -- file an issue about the crash?

**Review by costinm — COMMENTED** [review]
Re. testing: this depends on a separate PR in installer, with a new injection template and istiod updates.

The important part - that the new SDS doesn't affect existing workloads - is covered by the existing test suites.

File: security/pkg/nodeagent/sds/sdsservice.go
Comment (costinm) [reply security/pkg/nodeagent/sds/sdsservice.go]: I think John did already, he found it while testing istiod.

Comment (lei-tang) [reply security/pkg/nodeagent/sds/sdsservice.go]: Can you provide more details about "using 2 grpc stacks at the same time"?

Comment (howardjohn) [reply security/pkg/nodeagent/sds/sdsservice.go]: Opened https://github.com/envoyproxy/envoy/issues/8941

**Review by howardjohn — APPROVED** [review]

**Pull request merged**
